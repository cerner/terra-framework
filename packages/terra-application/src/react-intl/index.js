export { default } from 'react-intl';
export * from 'react-intl';
