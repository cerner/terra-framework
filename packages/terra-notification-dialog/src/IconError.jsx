/* eslint-disable */
import React from 'react';

const IconError = () => {
  return (
    <svg viewBox="0 0 48 48" xmlns="http://www.w3.org/2000/svg">
      <path d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z" ></path><path fill="#FFF" d="M7 21.5h34v5H7v-5z" ></path>
    </svg>
  );
};

IconError.displayName = "IconError";
export default IconError;