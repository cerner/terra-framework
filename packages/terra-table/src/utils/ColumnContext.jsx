import React from 'react';

const ColumnContext = React.createContext({ pinnedColumnOffsets: [] , pinnedColumnHeaderOffsets: []});

export default ColumnContext;
