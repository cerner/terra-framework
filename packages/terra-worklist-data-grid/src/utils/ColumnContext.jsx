import React from 'react';

const ColumnContext = React.createContext();

export default ColumnContext;
