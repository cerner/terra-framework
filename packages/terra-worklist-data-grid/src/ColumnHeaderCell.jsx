import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import { injectIntl } from 'react-intl';
import ThemeContext from 'terra-theme-context';
import IconUp from 'terra-icon/lib/icon/IconUp';
import IconDown from 'terra-icon/lib/icon/IconDown';
import IconError from 'terra-icon/lib/icon/IconError';
import WorklistDataGridPropTypes from './proptypes/WorklistDataGridPropTypes';
import './_elementPolyfill';
import styles from './ColumnHeaderCell.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Data for columns. By default, columns will be presented in the order given.
   */
  column: WorklistDataGridPropTypes.columnShape.isRequired,
  /**
   * String that specifies the default width for columns in the grid. Any valid CSS width value is accepted.
   */
  width: PropTypes.string.isRequired,
  /**
   * String that specifies the column height. Any valid CSS height value accepted.
  */
  headerHeight: PropTypes.string.isRequired,
  /**
   * Function that is called when a selectable header cell is selected. Parameters: `onColumnSelect(columnId)`
   */
  onColumnSelect: PropTypes.func,
  /**
   * @private
   * Object containing intl APIs
   */
  intl: PropTypes.shape({ formatMessage: PropTypes.func }),
};

const ColumnHeaderCell = (props) => {
  const {
    column,
    width,
    headerHeight,
    onColumnSelect,
    intl,
  } = props;

  // Handle column header selection
  const onHeaderSelect = () => {
    onColumnSelect(column.id);
  };

  let sortIndicatorIcon;
  let errorIcon;

  // Add error icon when column error exists
  if (column.hasError) {
    errorIcon = <IconError a11yLabel={intl.formatMessage({ id: 'Terra.worklistDataGrid.columnError' })} className={cx('error-icon')} />;
  }

  // Add the sort indicator based on the sort direction
  if (column.sortIndicator === WorklistDataGridPropTypes.SortIndicators.ASCENDING) {
    sortIndicatorIcon = <IconUp />;
  } else if (column.sortIndicator === WorklistDataGridPropTypes.SortIndicators.DESCENDING) {
    sortIndicatorIcon = <IconDown />;
  }

  // Retrieve current theme from context
  const theme = useContext(ThemeContext);

  return (
  /* eslint-disable react/forbid-dom-props */
    <th
      key={column.id}
      className={cx('column-header', theme.className, { selectable: !(column.isSelectable === false) })}
      tabIndex="-1"
      role="columnheader"
      scope="col"
      aria-sort={column.sortIndicator}
      onClick={(!(column.isSelectable === false) && onColumnSelect) ? onHeaderSelect : undefined}
      style={{ width, height: headerHeight }}
    >
      <div className={cx('header-container')}>
        {errorIcon}
        <span role="button">{column.displayName}</span>
        {sortIndicatorIcon}
      </div>
    </th>
  );
};

ColumnHeaderCell.propTypes = propTypes;
export default injectIntl(ColumnHeaderCell);
