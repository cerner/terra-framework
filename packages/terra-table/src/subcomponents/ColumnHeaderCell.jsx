/* eslint-disable react-hooks/exhaustive-deps */
import React, {
  useCallback,
  useContext,
  useEffect,
  useRef,
} from 'react';
import PropTypes from 'prop-types';
import { injectIntl } from 'react-intl';
import * as KeyCode from 'keycode-js';
import classNames from 'classnames/bind';
import ThemeContext from 'terra-theme-context';
import VisuallyHiddenText from 'terra-visually-hidden-text';
import Button from 'terra-button';
import { IconUp, IconDown, IconError } from 'terra-icon';
import { validateAction } from '../proptypes/validators';
import ColumnResizeHandle from './ColumnResizeHandle';
import GridContext, { GridConstants } from '../utils/GridContext';
import { SortIndicators } from '../proptypes/columnShape';
import ColumnContext from '../utils/ColumnContext';
import styles from './ColumnHeaderCell.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Required string representing a unique identifier for the column header cell.
   */
  id: PropTypes.string.isRequired,

  /**
   * Unique identifier for the parent table
   */
  tableId: PropTypes.string.isRequired,

  /**
   * Unique identifier for the column
   */
  columnId: PropTypes.string.isRequired,

  /**
   * String of text to render within the column header cell.
   */
  displayName: PropTypes.string,

  /**
   * A string indicating which sorting indicator should be rendered. If not provided, no sorting indicator will be rendered.
   * If a `component` value is specified, `sortIndicator` will be ignored. One of `ascending`, `descending`.
   */
  sortIndicator: PropTypes.oneOf(Object.values(SortIndicators)),

  /**
   * Boolean value indicating whether or not the column has an error in the data.
   */
  hasError: PropTypes.bool,

  /**
   * Number that specifies the minimum column width in pixels.
   */
  minimumWidth: PropTypes.number,

  /**
   * Number that specifies the maximum column width in pixels.
   */
  maximumWidth: PropTypes.number,

  /**
   * Boolean value indicating whether or not the header cell is focused.
   */
  isActive: PropTypes.bool,

  /**
   * Boolean value indicating whether or not the header cell text is displayed in the cell.
   */
  isDisplayVisible: PropTypes.bool,

  /**
   * Boolean value indicating whether or not the column header is selectable.
   */
  isSelectable: PropTypes.bool,

  /**
   * Boolean value indicating whether or not the column header cell is an action cell.
   * The action cell might be a placeholder cell without actual action button
   */
  isActionCell: PropTypes.bool,

  /**
   * Data for action cell.
   */
  action: validateAction,

  /**
   * Boolean value indicating whether or not the column header is resizable.
   */
  isResizable: PropTypes.bool,

  /**
   * Boolean value indicating whether or not the column resize handle is active.
   */
  isResizeHandleActive: PropTypes.bool,

  /**
   * A function to be executed upon the resize handler activation to pass its data to parent component.
   * @param {element} leftNeighborCell - `columnHeaderCellRef.current`
   * Skip both parameters to indicate that there is no active resize handle at the moment.
   */
  resizeHandleStateSetter: PropTypes.func,

  /**
   * String that specifies the init height for the resize handler to accomodate actions row.
   */
  resizeHandlerInitHeight: PropTypes.string,

  /**
   * A `ref current` for neighbouring the active resize handle element, which needs to get focus on left arrow key event.
   */
  activeResizeHandlerNeighborCell: PropTypes.object,

  /**
   * Height of the parent table.
   */
  tableHeight: PropTypes.number,

  /**
   * Boolean value indicating whether or not the column header is resizable.
   */
  isResizeActive: PropTypes.bool,

  /**
   * Numeric increment in pixels to adjust column width when resizing via the keyboard.
   */
  columnResizeIncrement: PropTypes.number,

  /**
   * String that specifies the default width for columns in the grid. Any valid CSS width value is accepted.
   */
  width: PropTypes.number.isRequired,

  /**
   * String that specifies the column height. Any valid CSS height value accepted.
  */
  headerHeight: PropTypes.string.isRequired,

  /**
   * The cell's column position in the grid. This is zero based.
   */
  columnIndex: PropTypes.number,

  /**
   * Function that is called when a selectable header cell is selected. Parameters:
   * @param {string} rowId rowId
   * @param {string} columnId columnId
   */
  onColumnSelect: PropTypes.func,

  /**
   * Function that is called when the mouse down event is triggered on the column resize handle.
   */
  onResizeMouseDown: PropTypes.func,

  /**
   * Function that is called when the the keyboard is used to adjust the column size.
   */
  onResizeHandleChange: PropTypes.func,

  /**
   * @private
   * Object containing intl APIs
   */
  intl: PropTypes.shape({ formatMessage: PropTypes.func }),
};

const defaultProps = {
  hasError: false,
  isSelectable: false,
  isActive: false,
  isDisplayVisible: true,
  isResizable: false,
  isResizeActive: false,
};

const ColumnHeaderCell = (props) => {
  const {
    id,
    tableId,
    isActionCell,
    action,
    displayName,
    sortIndicator,
    hasError,
    isActive,
    isDisplayVisible,
    isSelectable,
    isResizable,
    isResizeHandleActive,
    resizeHandleStateSetter,
    resizeHandlerInitHeight,
    activeResizeHandlerNeighborCell,
    columnId,
    tableHeight,
    isResizeActive,
    columnResizeIncrement,
    width,
    minimumWidth,
    maximumWidth,
    headerHeight,
    onColumnSelect,
    intl,
    columnIndex,
    onResizeMouseDown,
    onResizeHandleChange,
  } = props;

  const columnContext = useContext(ColumnContext);
  const gridContext = useContext(GridContext);

  const columnHeaderCellRef = useRef();

  const setResizeHandleActive = useCallback((setActive) => {
    if (setActive) {
      resizeHandleStateSetter(columnId, columnHeaderCellRef?.current);
    } else {
      resizeHandleStateSetter();
    }
  }, [columnId, resizeHandleStateSetter]);

  const isGridContext = gridContext.role === GridConstants.GRID;

  useEffect(() => {
    if (isActive) {
      if (isResizable && isResizeActive) {
        setResizeHandleActive(true);
      } else {
        columnHeaderCellRef.current.focus();
        setResizeHandleActive(false);
      }
    } else {
      setResizeHandleActive(false);
    }
  }, [isActive, isResizable, isResizeActive]);

  const onResizeHandleMouseDown = useCallback((event) => {
    event.stopPropagation();
    if (onResizeMouseDown) {
      onResizeMouseDown(event, columnIndex, columnHeaderCellRef.current.offsetWidth);
    }
  }, [columnIndex, onResizeMouseDown]);

  // Restore focus to column header after resize action is completed.
  const onResizeHandleMouseUp = useCallback(() => {
    columnHeaderCellRef.current.focus();
    setResizeHandleActive(false);
  }, []);

  // Handle column header selection via the mouse click.
  const handleMouseDown = () => {
    onColumnSelect(columnId);
  };

  // Handle column header selection via the space bar.
  const handleKeyDown = (event, callback) => {
    const key = event.keyCode;
    switch (key) {
      case KeyCode.KEY_SPACE:
      case KeyCode.KEY_RETURN:
        if (callback) {
          // for action button
          callback();
          break;
        } else if (isSelectable && onColumnSelect) {
          onColumnSelect(columnId);
        }
        event.stopPropagation();
        event.preventDefault(); // prevent the default scrolling
        break;
      case KeyCode.KEY_LEFT:
        if (isResizable && isResizeHandleActive && isGridContext) {
          activeResizeHandlerNeighborCell?.focus();
          setResizeHandleActive(false);
          event.stopPropagation();
          event.preventDefault();
        }
        break;
      case KeyCode.KEY_RIGHT:
        if (isResizable && !isResizeHandleActive && isGridContext) {
          setResizeHandleActive(true);
          event.stopPropagation();
          event.preventDefault();
        }
        break;
      default:
    }
  };

  const errorIcon = hasError && <IconError className={cx('error-icon')} />;

  // Add the sort indicator based on the sort direction
  let sortIndicatorIcon;
  let sortDescription = '';
  if (sortIndicator === SortIndicators.ASCENDING) {
    sortIndicatorIcon = <IconUp />;
    sortDescription = intl.formatMessage({ id: 'Terra.table.sort-ascending' });
  } else if (sortIndicator === SortIndicators.DESCENDING) {
    sortIndicatorIcon = <IconDown />;
    sortDescription = intl.formatMessage({ id: 'Terra.table.sort-descending' });
  }

  // Retrieve current theme from context
  const theme = useContext(ThemeContext);

  // Calculate cell left position for pinned columns due to their sticky position style
  const cellLeftEdge = (columnIndex < columnContext.pinnedColumnOffsets.length) ? columnContext.pinnedColumnOffsets[columnIndex] : null;

  // For tables, we want elements to be tabbable when selectable, but not anytime else.
  let buttonTabIndex = isSelectable ? 0 : undefined;
  if (isGridContext) {
    // For grids, we only want 1 tab stop. We then define the focus behavior in DataGrid.
    buttonTabIndex = isSelectable && displayName ? -1 : undefined;
  }

  // Determine if button element is required for column header
  const hasButtonElement = isSelectable && displayName;

  // Format header description for screenreader
  let headerDescription = displayName;
  headerDescription += errorIcon ? `, ${intl.formatMessage({ id: 'Terra.table.columnError' })}` : '';
  headerDescription += sortDescription ? `, ${sortDescription}` : '';

  // set focus in action cell to either button or resize handler
  /*
  const distributeFocusWithinActionCell = () => {
    columnHeaderCellRef?.current?.childNodes[0].focus();
    columnHeaderCellRef?.current?.childNodes[0].setAttribute('aria-hidden', false);
  };

  const onBlur = () => {
    columnHeaderCellRef?.current?.childNodes[0].setAttribute('aria-hidden', true);
  }; */

  const isPinnedColumn = columnIndex < columnContext.pinnedColumnOffsets.length;
  const CellTag = !isActionCell ? 'th' : 'td';

  // Create cell content
  let cellContent;
  if (isActionCell) {
    if (action) {
      cellContent = (
        <Button
          variant="de-emphasis"
          ref={columnHeaderCellRef}
          isCompact
          onClick={action.onClick}
          onKeyDown={(event) => handleKeyDown(event, action?.onClick)}
          text={action.label}
        />
      );
    } else {
      cellContent = (
        <span className={cx('display-text', 'hidden')}>
          {intl.formatMessage({ id: 'Terra.table.noAction' })}
        </span>
      );
    }
  } else {
    cellContent = (
      <div
        className={cx('header-container')}
        {...hasButtonElement && { ref: columnHeaderCellRef, role: 'button' }}
        tabIndex={buttonTabIndex}
      >
        {errorIcon}
        <span aria-hidden className={cx('display-text', { hidden: !isDisplayVisible })}>{displayName}</span>
        {sortIndicatorIcon}
        <VisuallyHiddenText text={headerDescription} />
      </div>
    );
  }

  return (
  /* eslint-disable react/forbid-dom-props */
    <CellTag
      ref={!hasButtonElement ? columnHeaderCellRef : undefined}
      id={`${tableId}-${id}`}
      key={id}
      className={cx('column-header', theme.className, {
        'action-cell': isActionCell,
        selectable: isSelectable,
        pinned: isPinnedColumn,
        'last-pinned-column': columnIndex === columnContext.pinnedColumnOffsets.length - 1,
      })}
      tabIndex={isGridContext && !hasButtonElement ? -1 : undefined}
      role={!isActionCell ? 'columnheader' : undefined}
      scope={!isActionCell ? 'col' : undefined}
      aria-owns={!isActionCell ? `${columnId}-actionCell` : undefined}
      title={!isActionCell ? displayName : action?.label}
      onMouseDown={isSelectable && onColumnSelect ? handleMouseDown : undefined}
      onKeyDown={(isSelectable || isResizable) ? handleKeyDown : undefined}
      // eslint-disable-next-line react/forbid-component-props
      style={{ width: `${width}px`, height: isActionCell ? 'auto' : headerHeight, left: cellLeftEdge }}
      // onFocus={isActionCell ? distributeFocusWithinActionCell : undefined}
      // onBlur={isActionCell ? onBlur : undefined}
    >
      {cellContent}
      { isResizable && !isActionCell && (
      <ColumnResizeHandle
        columnIndex={columnIndex}
        columnText={displayName}
        columnWidth={width}
        columnResizeIncrement={columnResizeIncrement}
        isActive={isResizeHandleActive}
        setIsActive={setResizeHandleActive}
        height={tableHeight}
        resizeHandlerInitHeight={resizeHandlerInitHeight}
        minimumWidth={minimumWidth}
        maximumWidth={maximumWidth}
        onResizeMouseDown={onResizeHandleMouseDown}
        onResizeMouseUp={onResizeHandleMouseUp}
        onResizeHandleChange={onResizeHandleChange}
      />
      )}
    </CellTag>
  );
};

ColumnHeaderCell.propTypes = propTypes;
ColumnHeaderCell.defaultProps = defaultProps;
export default React.memo(injectIntl(ColumnHeaderCell));
