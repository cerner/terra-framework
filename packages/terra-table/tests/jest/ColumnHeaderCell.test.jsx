import React from 'react';
import IconUp from 'terra-icon/lib/icon/IconUp';
import IconDown from 'terra-icon/lib/icon/IconDown';
import IconError from 'terra-icon/lib/icon/IconError';
import { IntlProvider } from 'react-intl';
import ColumnHeaderCell from '../../src/subcomponents/ColumnHeaderCell';
import ColumnContext from '../../src/utils/ColumnContext';

beforeAll(() => {
  jest.spyOn(console, 'error').mockImplementation();
  jest.spyOn(console, 'warn').mockImplementation();
});

afterAll(() => {
  console.error.mockRestore(); // eslint-disable-line no-console
  console.warn.mockRestore(); // eslint-disable-line no-console
});

describe('ColumnHeaderCell', () => {
  it('renders a default column header cell', () => {
    const column = {
      id: 'Column-0',
      displayName: 'Vitals',
      isSelectable: true,
    };

    const wrapper = enzymeIntl.mountWithIntl(
      <IntlProvider locale="en">
        <ColumnHeaderCell
          tableId="test-table"
          columnIndex={0}
          width={100}
          headerHeight="150px"
          {...column}
        />
      </IntlProvider>,
    );

    const columnHeader = wrapper.find('.column-header.selectable');
    expect(columnHeader).toHaveLength(1);
    expect(columnHeader.key()).toBe('Column-0');
    expect(columnHeader.props().id).toBe('test-table-Column-0');
    expect(columnHeader.props().role).toBe('columnheader');
    expect(columnHeader.props().scope).toBe('col');
    expect(columnHeader.props().tabIndex).toEqual(undefined);
    expect(columnHeader.props().style.width).toBe('100px');
    expect(columnHeader.props().style.height).toBe('150px');
    expect(columnHeader.props().title).toBe('Vitals');

    const headerContainer = columnHeader.find('.header-container[role="button"]');
    expect(headerContainer.find('.display-text').text().trim()).toBe('Vitals');
    expect(headerContainer.find('VisuallyHiddenText').text()).toBe('Vitals');
  });

  it('renders a column header cell with ascending sort', () => {
    const column = {
      id: 'Column-0',
      displayName: 'Vitals',
      sortIndicator: 'ascending',
      isSelectable: true,
    };

    const wrapper = enzymeIntl.mountWithIntl(
      <IntlProvider locale="en">
        <ColumnHeaderCell
          tableId="test-table"
          columnIndex={0}
          width={100}
          headerHeight="150px"
          {...column}
        />
      </IntlProvider>,
    );

    const columnHeader = wrapper.find('.column-header.selectable');
    expect(columnHeader).toHaveLength(1);
    expect(columnHeader.key()).toBe('Column-0');
    expect(columnHeader.props().id).toBe('test-table-Column-0');
    expect(columnHeader.props().role).toBe('columnheader');
    expect(columnHeader.props().scope).toBe('col');
    expect(columnHeader.props().tabIndex).toEqual(undefined);
    expect(columnHeader.props().style.width).toBe('100px');
    expect(columnHeader.props().style.height).toBe('150px');
    expect(columnHeader.props().title).toBe('Vitals');

    const headerContainer = columnHeader.find('.header-container[role="button"]');
    expect(headerContainer.find('.display-text').text().trim()).toBe('Vitals');
    expect(headerContainer.find(IconUp)).toHaveLength(1);
    expect(headerContainer.find('VisuallyHiddenText').text()).toBe('Vitals, Terra.table.sort-ascending');
  });

  it('renders a column header cell with descending sort', () => {
    const column = {
      id: 'Column-0',
      displayName: 'Vitals',
      sortIndicator: 'descending',
      isSelectable: true,
    };

    const wrapper = enzymeIntl.mountWithIntl(
      <IntlProvider locale="en">
        <ColumnHeaderCell
          tableId="test-table"
          columnIndex={0}
          width={100}
          headerHeight="150px"
          {...column}
        />
      </IntlProvider>,
    );

    const columnHeader = wrapper.find('.column-header.selectable');
    expect(columnHeader).toHaveLength(1);
    expect(columnHeader.key()).toBe('Column-0');
    expect(columnHeader.props().id).toBe('test-table-Column-0');
    expect(columnHeader.props().role).toBe('columnheader');
    expect(columnHeader.props().scope).toBe('col');
    expect(columnHeader.props().tabIndex).toEqual(undefined);
    expect(columnHeader.props().style.width).toBe('100px');
    expect(columnHeader.props().style.height).toBe('150px');
    expect(columnHeader.props().title).toBe('Vitals');

    const headerContainer = columnHeader.find('.header-container[role="button"]');
    expect(headerContainer.find('.display-text').text().trim()).toBe('Vitals');
    expect(headerContainer.find(IconDown)).toHaveLength(1);
    expect(headerContainer.find('VisuallyHiddenText').text()).toBe('Vitals, Terra.table.sort-descending');
  });

  it('renders a column header cell with error', () => {
    const column = {
      id: 'Column-0',
      displayName: 'Vitals',
      hasError: true,
      isSelectable: true,
    };

    const wrapper = enzymeIntl.mountWithIntl(
      <IntlProvider locale="en">
        <ColumnHeaderCell
          tableId="test-table"
          columnIndex={0}
          width={100}
          headerHeight="150px"
          {...column}
        />
      </IntlProvider>,
    );

    const columnHeader = wrapper.find('.column-header.selectable');
    expect(columnHeader).toHaveLength(1);
    expect(columnHeader.key()).toBe('Column-0');
    expect(columnHeader.props().id).toBe('test-table-Column-0');
    expect(columnHeader.props().role).toBe('columnheader');
    expect(columnHeader.props().scope).toBe('col');
    expect(columnHeader.props().tabIndex).toEqual(undefined);
    expect(columnHeader.props().style.width).toBe('100px');
    expect(columnHeader.props().style.height).toBe('150px');
    expect(columnHeader.props().title).toBe('Vitals');

    const headerContainer = columnHeader.find('.header-container[role="button"]');
    expect(headerContainer.find('.display-text').text().trim()).toBe('Vitals');
    expect(headerContainer.find(IconError)).toHaveLength(1);
    expect(headerContainer.find('VisuallyHiddenText').text()).toBe('Vitals, Terra.table.columnError');
  });

  it('renders a column header cell with ascending sort and error', () => {
    const column = {
      id: 'Column-0',
      displayName: 'Vitals',
      sortIndicator: 'ascending',
      hasError: true,
      isSelectable: true,
    };

    const wrapper = enzymeIntl.mountWithIntl(
      <IntlProvider locale="en">
        <ColumnHeaderCell
          tableId="test-table"
          columnIndex={0}
          width={100}
          headerHeight="150px"
          {...column}
        />
      </IntlProvider>,
    );

    const columnHeader = wrapper.find('.column-header.selectable');
    expect(columnHeader).toHaveLength(1);
    expect(columnHeader.key()).toBe('Column-0');
    expect(columnHeader.props().id).toBe('test-table-Column-0');
    expect(columnHeader.props().role).toBe('columnheader');
    expect(columnHeader.props().scope).toBe('col');
    expect(columnHeader.props().tabIndex).toEqual(undefined);
    expect(columnHeader.props().style.width).toBe('100px');
    expect(columnHeader.props().style.height).toBe('150px');
    expect(columnHeader.props().title).toBe('Vitals');

    const headerContainer = columnHeader.find('.header-container[role="button"]');
    expect(headerContainer.find('.display-text').text().trim()).toBe('Vitals');
    expect(headerContainer.find(IconUp)).toHaveLength(1);
    expect(headerContainer.find(IconError)).toHaveLength(1);
    expect(headerContainer.find('VisuallyHiddenText').text()).toBe('Vitals, Terra.table.columnError, Terra.table.sort-ascending');
  });

  it('renders a column header cell with onColumnSelect callback', () => {
    const column = {
      id: 'Column-0',
      displayName: 'Vitals',
      sortIndicator: 'ascending',
      hasError: true,
      isSelectable: true,
    };

    const mockClick = jest.fn();

    const wrapper = enzymeIntl.mountWithIntl(
      <IntlProvider locale="en">
        <ColumnHeaderCell
          tableId="test-table"
          columnIndex={0}
          width={100}
          headerHeight="150px"
          {...column}
          onColumnSelect={mockClick}
        />
      </IntlProvider>,
    );

    const columnHeader = wrapper.find('.column-header.selectable');
    expect(columnHeader).toHaveLength(1);
    expect(columnHeader.key()).toBe('Column-0');
    expect(columnHeader.props().id).toBe('test-table-Column-0');
    expect(columnHeader.props().role).toBe('columnheader');
    expect(columnHeader.props().scope).toBe('col');
    expect(columnHeader.props().tabIndex).toEqual(undefined);
    expect(columnHeader.props().onMouseDown).toBeDefined();
    expect(columnHeader.props().style.width).toBe('100px');
    expect(columnHeader.props().style.height).toBe('150px');
    expect(columnHeader.props().title).toBe('Vitals');

    const headerContainer = columnHeader.find('.header-container[role="button"]');
    expect(headerContainer.props().tabIndex).toBe(0);
    expect(headerContainer.find('.display-text').text().trim()).toBe('Vitals');
    expect(headerContainer.find(IconUp)).toHaveLength(1);
    expect(headerContainer.find(IconError)).toHaveLength(1);
    expect(headerContainer.find('VisuallyHiddenText').text()).toBe('Vitals, Terra.table.columnError, Terra.table.sort-ascending');
  });

  it('renders a column header cell with onColumnSelect callback but not selectable', () => {
    const column = {
      id: 'Column-0',
      displayName: 'Vitals',
      sortIndicator: 'ascending',
      hasError: true,
      isSelectable: false,
    };

    const mockColumnSelect = jest.fn();

    const wrapper = enzymeIntl.mountWithIntl(
      <IntlProvider locale="en">
        <ColumnHeaderCell
          tableId="test-table"
          columnIndex={0}
          width={100}
          headerHeight="150px"
          {...column}
          onColumnSelect={mockColumnSelect}
        />
      </IntlProvider>,
    );

    const columnHeader = wrapper.find('.column-header:not(selectable)');
    expect(columnHeader).toHaveLength(1);
    expect(columnHeader.key()).toBe('Column-0');
    expect(columnHeader.props().id).toBe('test-table-Column-0');
    expect(columnHeader.props().role).toBe('columnheader');
    expect(columnHeader.props().scope).toBe('col');
    expect(columnHeader.props().tabIndex).toEqual(undefined);
    expect(columnHeader.props().onMouseDown).toBeUndefined();
    expect(columnHeader.props().style.width).toBe('100px');
    expect(columnHeader.props().style.height).toBe('150px');
    expect(columnHeader.props().title).toBe('Vitals');

    const headerContainer = columnHeader.find('.header-container');
    expect(headerContainer.props().tabIndex).toBeUndefined();
    expect(headerContainer.find('.display-text').text().trim()).toBe('Vitals');
    expect(headerContainer.find(IconUp)).toHaveLength(1);
    expect(headerContainer.find(IconError)).toHaveLength(1);
    expect(headerContainer.find('VisuallyHiddenText').text()).toBe('Vitals, Terra.table.columnError, Terra.table.sort-ascending');
  });

  it('renders a pinned column header cell', () => {
    const column = {
      id: 'Column-0',
      displayName: 'Vitals',
      sortIndicator: 'ascending',
      hasError: true,
    };

    const wrapper = enzymeIntl.mountWithIntl(
      <ColumnContext.Provider value={{ pinnedColumnOffsets: [0] }}>
        <ColumnHeaderCell
          tableId="test-table"
          columnIndex={0}
          width={100}
          headerHeight="150px"
          {...column}
        />
      </ColumnContext.Provider>,
    );

    expect(wrapper.find('.pinned')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls a custom column select callback function on mouse down', () => {
    const mockOnColumnSelect = jest.fn();
    const wrapper = enzymeIntl.mountWithIntl(
      <ColumnHeaderCell
        onColumnSelect={mockOnColumnSelect}
        isSelectable
      />,
    );
    wrapper.find('.column-header').simulate('mousedown');

    // Validate mock function was called from simulated onMouseDown event
    expect(mockOnColumnSelect).toHaveBeenCalled();
  });

  it('verifes that the custom column select callback is not called on mouse down when not selectable', () => {
    const mockOnColumnSelect = jest.fn();
    const wrapper = enzymeIntl.mountWithIntl(
      <ColumnHeaderCell
        onColumnSelect={mockOnColumnSelect}
      />,
    );
    wrapper.find('.column-header').simulate('mousedown');

    // Validate mock function was called from simulated onMouseDown event
    expect(mockOnColumnSelect).not.toHaveBeenCalled();
  });
});
