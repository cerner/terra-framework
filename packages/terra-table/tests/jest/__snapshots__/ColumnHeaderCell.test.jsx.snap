// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnHeaderCell renders a pinned column header cell 1`] = `
<InjectIntl(ColumnHeaderCell)
  columnIndex={0}
  displayName="Vitals"
  hasError={true}
  headerHeight="150px"
  id="Column-0"
  sortIndicator="ascending"
  tableId="test-table"
  width={100}
>
  <ColumnHeaderCell
    columnIndex={0}
    displayName="Vitals"
    hasError={true}
    headerHeight="150px"
    id="Column-0"
    intl={
      Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": null,
        "now": [Function],
        "onError": [Function],
        "textComponent": "span",
        "timeZone": null,
      }
    }
    isActive={false}
    isDisplayVisible={true}
    isResizable={false}
    isResizeActive={false}
    isSelectable={false}
    sortIndicator="ascending"
    tableId="test-table"
    width={100}
  >
    <th
      className="column-header pinned"
      id="test-table-Column-0"
      key="Column-0"
      onKeyDown={[Function]}
      role="columnheader"
      scope="col"
      style={
        Object {
          "height": "150px",
          "left": 0,
          "width": "100px",
        }
      }
      title="Vitals"
    >
      <div
        className="header-container"
      >
        <IconError
          className="error-icon"
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        >
          <IconBase
            className="IconError error-icon"
            focusable="false"
            height="1em"
            isBidi={false}
            isSpin={false}
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <svg
              className="tui-Icon icon IconError error-icon"
              focusable="false"
              height="1em"
              role="presentation"
              viewBox="0 0 48 48"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 0c13.3 0 24 10.7 24 24S37.3 48 24 48 0 37.3 0 24C0 10.8 10.7 0 23.9 0h.1z"
                fill="#E50000"
              />
              <path
                d="M7 21.5h34v5H7v-5z"
                fill="#FFF"
              />
            </svg>
          </IconBase>
        </IconError>
        <span
          aria-hidden={true}
          className="display-text"
        >
          Vitals
        </span>
        <IconUp
          data-name="Layer 1"
          viewBox="0 0 48 48"
          xmlns="http://www.w3.org/2000/svg"
        >
          <IconBase
            data-name="Layer 1"
            focusable="false"
            height="1em"
            isBidi={false}
            isSpin={false}
            viewBox="0 0 48 48"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <svg
              className="tui-Icon icon"
              data-name="Layer 1"
              focusable="false"
              height="1em"
              role="presentation"
              viewBox="0 0 48 48"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 0 .1 23.5l3.8 3.8L21 10.6V48h6V10.6l17.1 16.7 3.8-3.8z"
              />
            </svg>
          </IconBase>
        </IconUp>
        <VisuallyHiddenText
          text="Vitals, Terra.table.columnError, Terra.table.sort-ascending"
        >
          <span
            className="visually-hidden-text"
          >
            Vitals, Terra.table.columnError, Terra.table.sort-ascending
          </span>
        </VisuallyHiddenText>
      </div>
      <div
        className="pinned-columns-divider"
        style={
          Object {
            "height": undefined,
            "left": 99,
          }
        }
      />
    </th>
  </ColumnHeaderCell>
</InjectIntl(ColumnHeaderCell)>
`;
